File: algorithms/valid_paren/valid_paren.cpp
/*
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * An input string is valid if:
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Note that an empty string is also considered valid. 
 */

class ValidParen {
  public:
    bool isValid(string s) {
    stack<char> stack;

    if(s.length() == 0){
        return true;
    }
    if(s.length() == 1){
        return false;
    }
    
    for(int i =0 ; i < s.length(); i++){
        char c = s[i];
        if(c == '}'|| c == ')' || c == ']'){
            if(stack.size() == 0 ) return false;
            if(c == '}' && stack.top() != '{') return false;
            if(c == ')' && stack.top() != '(') return false;
            if(c == ']' && stack.top() != '[') return false;
            stack.pop();
            
         } else {
            stack.push(c);
        }
    }
    if(stack.size() == 0) return true;
    return false;
    }
};

File: algorithms/valid_paren/valid_paren.py
# Valid Parenthesis Problem 

def valid_paren(parens):
    stack = []
    parens = list(parens)
    for i in range(len(parens)):
        current = parens[i]
        if (current == '[' or current == '(' or current == '{'):
            stack.append(current)
        elif (current == "}"):
            if (stack.pop() != '{'):
                return False
        elif (current == ")"):
            if (stack.pop() != '('):
                return False
        elif (current != "["):
            if (stack.pop() != '['):
                return False
    return True   

def tests():
    assert(valid_paren("()") == True)
    print(1)
    assert(valid_paren("()[]{}") == True)
    print(2)
    assert(valid_paren("(]") == False)
    print(3)
    assert(valid_paren("([)]") == False)
    print(4)
    return 0

tests()